cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -430,7 +433,33 @@ template <>
 hash_value_type CUDA_DEVICE_CALLABLE
 SparkMurmurHash3_32<numeric::decimal128>::operator()(numeric::decimal128 const& key) const
 {
-  return this->compute<__int128_t>(key.value());
+  // Generates the Spark murmur3 hash value, mimicing the conversion
+  // java.math.BigDecimal.valueOf(unscaled_value, _scale).unscaledValue().toByteArray()
+  // https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L381
+  __int128_t const val    = key.value();
+  int8_t const* data      = reinterpret_cast<int8_t const*>(&val);
+  int8_t const sign_bit   = data[15] & 0x80;
+  int8_t const zero_value = sign_bit ? 0xff : 0x00;
+  int32_t length          = 15;
+
+  // Search for first non-zero byte in the unscaled value, shortening the hashed data
+  for (; length >= 0 && data[length] == zero_value; --length) {}
+  // Special case for 0 which does not shorten correctly
+  if (length == -1) { return this->compute<uint8_t>(0); }
```

Comment:
The special case for zero should be handled before iterating over the bytes of the value if possible. This special case should only be reached if the `__int128` value is equal to zero (`0x00000000...`, right? Is this case also supposed to catch `0xffffffff...`? It seems like it would, if the sign bit is 1 and all the rest of the bytes in the value are `0xff`, but it's not clear if that's the intended behavior. If it is, we should indicate in a comment that both values are expected to be caught.

Reply:
I think that's a bug, double checking against Spark

Reply:
This was a bug, `-1` is now explicitly handled